File: src/dynamic-programming/maximum-length-of-repeated-subarray.cpp
class Solution {
public:
	long long myPow(long long x, int n, int mod) {
		long long ans = 1;
		while (n) {
			if (n & 1) {
				ans = ans * x % mod;
			}

			x = x * x % mod;
			n >>= 1;
		}

		return ans;
	}

	int check(vector<int>& nums1, vector<int>& nums2, int len) {
		if (!len) return false;
		const int mod = 1000000009;
		const int base = 113;
		const int n = nums1.size();
		const int m = nums2.size();
		long long p13331 = myPow(base, len - 1, mod);
		set<long long> hashBuckets1;

		long long hash1 = 0;
		for (int i = 0; i < len; ++i) {
			hash1 = (hash1 * base + nums1[i]) % mod;;
		}

		hashBuckets1.insert(hash1);

		for (int i = len; i < n; ++i) {
			hash1 = ((hash1 - nums1[i - len] * p13331 % mod + mod) % mod * base + nums1[i]) % mod;
			hashBuckets1.insert(hash1);
		}

		long long hash2 = 0;

		for (int i = 0; i < len; ++i) {
			hash2 = (hash2 * base + nums2[i]) % mod;
		}

		if (hashBuckets1.count(hash2)) return true;


		for (int i = len; i < m; ++i) {
			hash2 = ((hash2 - nums2[i - len] * p13331 % mod + mod) % mod * base + nums2[i]) % mod;
			if (hashBuckets1.count(hash2)) return true;
		}

		return false;
	}

	int findLength(vector<int>& nums1, vector<int>& nums2) {
		int l = 1;
		int r = min(nums1.size(), nums2.size());
		int ans = 0;
		while (l <= r) {
			int mid = l + ((r - l) >> 1);
			if (check(nums1, nums2, mid)) {
				ans = mid;
				l = mid + 1;
			}
			else {
				r = mid - 1;
			}
		}

		return ans;
	}
};

File: src/dynamic-programming/maximum-length-of-repeated-subarray.ts
//自己做题,复杂度O(n ^ 3)
export function findLength(nums1: number[], nums2: number[]): number {
  let ans = 0
  const n = nums1.length
  const m = nums2.length

  const getCommonArrayLength = (s1: number, s2: number): number => {
    let ans = 0
    while (s1 < n && s2 < m && nums1[s1] === nums2[s2]) {
      ++ans
      ++s1
      ++s2
    }

    return ans
  }

  for (let i = 0; i < n; ++i) {
    for (let j = 0; j < m; ++j) {
      ans = Math.max(getCommonArrayLength(i, j), ans)
    }
  }

  return ans
}

//动态规划解法，时间复杂度O(n * m)
export function findLength1(nums1: number[], nums2: number[]): number {
  const n = nums1.length
  const m = nums2.length
  //dp[i][j]表示在nums1中以nums1[i]元素为开头的子数组
  //和nums2中以nums2[j]为开头的子数组最大的公共前缀长度
  //A: [1,2,3,2,1]
  //B: [3,2,1,4,7]
  //就比如dp[3][1]就为2,因为数组[2,1]和数组[2,1,4,7]最大的公共前缀长度为2
  
  const dp: number[][] = Array.from({ length: n + 1 }, () =>
    Array.from<number>({ length: m + 1 }).fill(0)
  )

  let ans = 0
  for (let i = n - 1; i >= 0; --i) {
    for (let j = m - 1; j >= 0; --j) {
      dp[i][j] = nums1[i] === nums2[j] ? dp[i + 1][j + 1] + 1 : 0
      ans = Math.max(dp[i][j], ans)
    }
  }

  return ans
}
